### PR TITLE
Add istanbul.timeout to governance items

### PR DESCRIPTION
**Pull request by andybclee**
## Proposed changes

- This PR is to make the default timeout for a roundchange changeable by the governance vote
- Added an API `istanbul.roundchangeTimer` to check current timeout value in milliseconds
- Added a governance key of `istanbul.roundchangetimer` for `governance.vote` API
  ex) 
```shell
> istanbul.roundchangeTimer
3000
>governance.vote("istanbul.roundchangeTimer", 10000) 
==> this will change the timeout interval into 10 seconds
> istanbul.roundchangeTimer
10000
```
- This change is applied instantly as soon as the vote passes like `addValidator` and `removeValidator`
- Added test codes to test the new feature

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by kjhman21** [comment]
requesttimeout is too general. Do you want to use this parameter for other cases?

**Comment by andybclee** [comment]
> requesttimeout is too general. Do you want to use this parameter for other cases?

@kjhman21 It may seem so. But it is in the `istanbul` namespace.In the namespace `requesttimeout` has one meaning. So I think its name is okay.

**Comment by kjhman21** [comment]
@andybclee Even if it has 'istanbul' namespace, 'request' can have various meaning.
request of prepare message
Why don't you use 'round change timer'?

**Comment by andybclee** [comment]
> @andybclee Even if it has 'istanbul' namespace, 'request' can have various meaning.
> request of prepare message
> Why don't you use 'round change timer'?

Good suggestion. Thanks! 
I'll apply it right away.

File: governance/handler.go [hunk 1]
```diff
@@ -60,6 +60,7 @@ var GovernanceItems = map[int]check{
 	params.Policy:                  {uint64T, checkUint64andBool, updateProposerPolicy},
 	params.CommitteeSize:           {uint64T, checkUint64andBool, nil},
 	params.ConstTxGasHumanReadable: {uint64T, checkUint64andBool, updateTxGasHumanReadable},
+	params.RequestTimeout:          {uint64T, checkUint64andBool, nil},
```
Comment (laggu) [inline governance/handler.go]: hasn't been changed to roundChangeTimer :)

File: params/governance_params.go [hunk 2]
```diff
@@ -87,6 +87,7 @@ const (
 	ProposerRefreshInterval
 	ConstTxGasHumanReadable
 	CliqueEpoch
+	RequestTimeout
```
Comment (laggu) [inline params/governance_params.go]: hasn't been changed to roundChangeTimer :)

File: governance/handler.go [hunk 3]
```diff
@@ -383,6 +384,12 @@ func (gov *Governance) addNewVote(valset istanbul.ValidatorSet, votes []Governan
 				target := gVote.Value.(common.Address)
 				valset.RemoveValidator(target)
 				votes = gov.removeVotesFromRemovedNode(votes, target)
+			case params.RequestTimeout:
+				timeout := gVote.Value.(uint64)
```
Comment (laggu) [inline governance/handler.go]: would be better to change the name of variable to roundChangeTimer

File: consensus/istanbul/backend/api.go [hunk 4]
```diff
@@ -423,3 +423,7 @@ func (api *APIExtension) GetBlockWithConsensusInfoByHash(blockHash common.Hash)
 
 	return api.makeRPCOutput(block, proposer, committee, block.Transactions(), receipts), nil
 }
+
+func (api *API) GetRequestTimeout() uint64 {
+	return istanbul.DefaultConfig.RequestTimeout
```
Comment (laggu) [inline consensus/istanbul/backend/api.go]: function name and istanbul.DefaultConfig.RequestTimeout hasn't been changed to roundChangeTimer

**Comment by andybclee** [comment]
@laggu @kjhman21 I think the point is that which word is more intuitive for users not for developers.

The word 'roundchange' is not a common word though we are very familiar with it. It is only used in istanbul bft. Normal BFT uses the word 'view change'. 

But I think 'request timeout' can be more straightforward for usual system administrators and it means enough for what it does. 

My opinion is that it would be better to revert the change of the name to 'requesttimeout'

**Comment by kjhman21** [comment]
@andybclee 
`request` is a general term. 'view change' is also hard to understand to users.
I'm not insisting that 'roundchange' is the best idea, but 'request' is not good because it cannot describe the exact meaning of something.

'requesttimeout'
- Is it requesting a timeout value?
- What kinds of request does it describe?

If this value is used for all message timeouts during the consensus round, what about using just 'timeout'?

istanbul.timeout?

**Comment by andybclee** [comment]
@kjhman21 I didn't mean that 'view change' is easy to understand. 
Actually the value works in
- waiting for a proposal
- waiting for prepare messages
- waiting for commit messages

It is all about time out. I agree that 'istanbul.timeout' fits for its usage.

**Comment by kjhman21** [comment]
@andybclee Okay, then please change the name to 'istanbul.timeout'! :)

File: governance/default.go [hunk 5]
```diff
@@ -533,7 +535,7 @@ func (gov *Governance) updateChangeSet(vote GovernanceVote) bool {
 	case params.GovernanceMode, params.Ratio:
 		gov.changeSet.SetValue(GovernanceKeyMap[vote.Key], vote.Value.(string))
 		return true
-	case params.Epoch, params.StakeUpdateInterval, params.ProposerRefreshInterval, params.CommitteeSize, params.UnitPrice, params.ConstTxGasHumanReadable:
+	case params.Epoch, params.StakeUpdateInterval, params.ProposerRefreshInterval, params.CommitteeSize, params.UnitPrice, params.ConstTxGasHumanReadable, params.Timeout:
```
Comment (KimKyungup) [inline governance/default.go]: You can merge this and `Policy` like `ParseVoteValue` on line 507.

File: governance/handler.go [hunk 6]
```diff
@@ -383,6 +384,12 @@ func (gov *Governance) addNewVote(valset istanbul.ValidatorSet, votes []Governan
 				target := gVote.Value.(common.Address)
 				valset.RemoveValidator(target)
 				votes = gov.removeVotesFromRemovedNode(votes, target)
+			case params.Timeout:
+				timeout := gVote.Value.(uint64)
```
Comment (KimKyungup) [inline governance/handler.go]: This type assertion will always succeed?

File: governance/default_test.go [hunk 7]
```diff
@@ -625,6 +631,22 @@ func TestGovernance_HandleGovernanceVote_None_mode(t *testing.T) {
 	}
 	gov.voteMap.Clear()
 
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Test for "istanbul.requesttimeout" in "none" mode
```
Comment (KimKyungup) [inline governance/default_test.go]:
```suggestion
	// Test for "istanbul.timeout" in "none" mode
```

File: governance/default_test.go [hunk 8]
```diff
@@ -690,11 +712,36 @@ func TestGovernance_HandleGovernanceVote_Ballot_mode(t *testing.T) {
...
 	valSet, votes, tally = gov.HandleGovernanceVote(valSet, votes, tally, header, council[2], self)
 	if _, ok := gov.changeSet.items["governance.unitprice"]; !ok {
-		t.Errorf("Vote should be applied yet but it was not")
+		t.Errorf("Vote should be applied but it was not")
 	}
 
 	gov.RemoveVote("governance.unitprice", uint64(22000), blockCounter.Uint64())
 
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Test for "istanbul.roundchangetimer" in "ballot" mode
```
Comment (KimKyungup) [inline governance/default_test.go]:
```suggestion
	// Test for "istanbul. timeout" in "ballot" mode
```

File: governance/default_test.go [hunk 9]
```diff
@@ -625,6 +631,22 @@ func TestGovernance_HandleGovernanceVote_None_mode(t *testing.T) {
...
+	header.Number = blockCounter.Add(blockCounter, common.Big1)
+	header.BlockScore = common.Big1
+
+	newValue := istanbul.DefaultConfig.Timeout + uint64(10000)
+	gov.AddVote("istanbul.timeout", newValue)
+	header.Vote = gov.GetEncodedVote(proposer, blockCounter.Uint64())
+
+	gov.HandleGovernanceVote(valSet, votes, tally, header, proposer, self)
+
+	if istanbul.DefaultConfig.Timeout != newValue {
```
Comment (KimKyungup) [inline governance/default_test.go]:
```suggestion
	assert.NotEqual(t, istanbul.DefaultConfig.Timeout, newValue) 
```

File: governance/default_test.go [hunk 10]
```diff
@@ -690,11 +712,36 @@ func TestGovernance_HandleGovernanceVote_Ballot_mode(t *testing.T) {
...
+	newValue := istanbul.DefaultConfig.Timeout + uint64(10000)
+	gov.AddVote("istanbul.timeout", newValue)
+
+	header.Vote = gov.GetEncodedVote(council[0], blockCounter.Uint64())
+	valSet, votes, tally = gov.HandleGovernanceVote(valSet, votes, tally, header, council[0], self)
+
+	header.Vote = gov.GetEncodedVote(council[1], blockCounter.Uint64())
+	valSet, votes, tally = gov.HandleGovernanceVote(valSet, votes, tally, header, council[1], self)
+
+	if istanbul.DefaultConfig.Timeout == newValue {
```
Comment (KimKyungup) [inline governance/default_test.go]:
```suggestion
	assert.Equal(t, istanbul.DefaultConfig.Timeout, newValue)
```

File: governance/default_test.go [hunk 11]
```diff
@@ -690,11 +712,36 @@ func TestGovernance_HandleGovernanceVote_Ballot_mode(t *testing.T) {
...
+		t.Errorf("Vote shouldn't be applied yet but it was applied")
+	}
+
+	header.Vote = gov.GetEncodedVote(council[2], blockCounter.Uint64())
+	valSet, votes, tally = gov.HandleGovernanceVote(valSet, votes, tally, header, council[2], self)
+	if istanbul.DefaultConfig.Timeout != newValue {
```
Comment (KimKyungup) [inline governance/default_test.go]:
```suggestion
	assert.NotEqual(t, istanbul.DefaultConfig.Timeout, newValue)
```

File: governance/handler.go (hunk 6 above)
Comment (andybclee) [reply governance/handler.go]: @KimKyungup At this point, yes. Type checking is done before getting into this function.

**Comment by andybclee** [comment]
Applied all feedbacks. Please take another look :)

File: consensus/istanbul/core/core.go [hunk 12]
```diff
@@ -353,7 +353,7 @@ func (c *core) newRoundChangeTimer() {
 	c.stopTimer()
 
 	// set timeout based on the round number
-	timeout := time.Duration(c.config.RequestTimeout) * time.Millisecond
+	timeout := time.Duration(c.config.Timeout) * time.Millisecond
```
Comment (kjhman21) [inline consensus/istanbul/core/core.go]: Is it okay to use c.config?

Comment (andybclee) [reply consensus/istanbul/core/core.go]: @kjhman21 Thank you for pointing this issue. Now it is changed to use `DefaultConfig` and confirmed that timeout is changed by the governance.vote.

**Comment by andybclee** [comment]
@kjhman21 @KimKyungup I added atomic update/read for the variable. PTAL.

**Comment by kjhman21** [comment]
@andybclee let me know if this PR is ready to be merged.

**Pull request merged**